### ISSUE TITLE
Client tooltips cut off by game on Mac

### ISSUE THREAD

**Issue by DADtheMaggot** [runelite/runelite#11696]
**Describe the bug**
When using the GPU Plugin on Mac, the game covers the tooltips that pop up when hovering over runelite's plugins/options. 

**To Reproduce**
Hover over a runelite plugin/option with a long enough description to extend out of the side bar. With GPU off the whole tooltip is available, with GPU on it gets cut off. 

**Expected behavior**
I'd expect the tooltips to appear above the game so you can read the whole thing. 

**Screenshots**
<img width="1440" alt="Screen Shot 2020-05-24 at 11 22 04 PM" src="https://user-images.githubusercontent.com/47622191/82775343-71dc8280-9e15-11ea-8b37-d018d3156db0.png">

<img width="386" alt="Screen Shot 2020-05-24 at 11 20 28 PM" src="https://user-images.githubusercontent.com/47622191/82775335-6b4e0b00-9e15-11ea-9789-43b280dc888e.png">

**Environment (please complete the following information):**
 - OS: macOS Mojave 10.14.6
 - RuneLite version: 2.1.2
 - Launcher version: 2.1.2
(I'm not 100% that I was looking in the right place for the RL and Launcher versions, but they should be the most recent as of this post.)

**Comment by keyosk** [comment]
Can confirm the same behavior.

- OS: macOS Mojave 10.14.6 
- GPU 1: Radeon Pro 460 4 GB (It should be using this card)
- GPU 0: Intel HD Graphics 530 1536 MB (This card is available but shouldn't be used)
- RuneLite version: 1.6.19
- Launcher version 2.1.0
- Oldschool revision: Rev 189

**Comment by richardcane** [comment]
Putting the tooltip text in a div and setting a width for the div forces the tooltip to render with a max width and wrap the text.

`
if (!pluginConfig.getDescription().isEmpty())
{
	nameLabel.setToolTipText("<html><div width='265'>" + pluginConfig.getName() + ":<br>" + pluginConfig.getDescription() + "</div></html>");
}
`

Which causes tooltips to look like:

![image](https://user-images.githubusercontent.com/29371470/110853926-1dbf6380-82ac-11eb-9031-527ea9b5dd0d.png)
![image](https://user-images.githubusercontent.com/29371470/110853999-36c81480-82ac-11eb-8ce1-40a704a747f0.png)

**Comment by Intecpsp** [comment]
I'm actually seeing this with the GPU plugin turned off on MacOs Mojave 10.14.6.

![Screen Shot 2021-06-07 at 2 12 58 PM](https://user-images.githubusercontent.com/1164012/121068639-85e1f080-c79a-11eb-8701-335aa04f073c.png)

**Comment by LlemonDuck** [comment]
Some findings:

This is not caused by the Substance LAF in use by RuneLite, it occurs even when disabled:

<img width="320" alt="image" src="https://user-images.githubusercontent.com/1868974/132109807-51a06cce-db8b-40b0-89ba-82753259060f.png">

It also occurs even if the side panel container is element 0 within the frame (before the client panel):

<img width="286" alt="image" src="https://user-images.githubusercontent.com/1868974/132109874-ec93ff71-0695-4de6-afd8-aa2738a60a84.png">

There is slightly different behaviour based on whether the GPU plugin is enabled or not, but I wasn't able to capture footage of it. With the GPU plugin:
* Enabled, tooltips are always and immediately rendered under the game canvas
* Disabled, tooltips are rendered above the game canvas until the next frame is drawn by the client, which is drawn over the tooltip. 
Disabling the call to `((Client)client).setGameDrawingMode(2);` within ClientPanel prevented this, when the tooltip did not overlap the viewport (fixed classic).


<img width="353" alt="image" src="https://user-images.githubusercontent.com/1868974/132110076-efbb0c3d-5de8-439d-bac4-93539ce8600c.png">

I did find a related SO question about the tooltip behaviour with JOGL canvases, and an affiliated JOGL bug report:
* https://stackoverflow.com/questions/37001959/tooltip-hidden-behind-jogl-glcanvas-on-os-x-when-using-non-aqua-look-and-feel
* https://jogamp.org/bugzilla/show_bug.cgi?id=1306
The JOGAmp ticket is resolved, although the associated fix doesn't seem related to me.

I found in the Substance LAF's `ShadowPopupFactory` (the default one in use here), it does not install itself onto macOS:
```java
    public static void install() {
        if (!LookUtils.IS_OS_MAC) {
            PopupFactory factory = PopupFactory.getSharedInstance();
            if (!(factory instanceof ShadowPopupFactory)) {
                PopupFactory.setSharedInstance(new ShadowPopupFactory(factory));
            }
        }
    }
```

The LAF PopupFactory in use on macOS is actually the `com.apple.laf.ScreenPopupFactory`. I will poke around inside that another time, but that's all I have for now.

**Comment by Adam-** [comment]
> Disabling the call to ((Client)client).setGameDrawingMode(2); within ClientPanel prevented this, when the tooltip did not overlap the viewport (fixed classic).

On fixed classic this prevents the client from typically drawing more than just the viewport per-frame, so it would make sense it would not cause this behavior. If you were to do something in game to trigger a ui redraw though it still would. So this is unrelated.

**Comment by LlemonDuck** [comment]
During runtime the ScreenPopupFactory's `fActive` is set to false (which controls heavyweighting of tooltips), likely when the Substance LAF is applied, although I didn't identify specifically why.

I managed to get a reflection-based PopupFactory that forces heavyweight tooltips to be rendered by the LAF by invoking `setPopupType` during `getPopup`. With this, I was able to render tooltips overtop of the game canvas. This worked both with the GPU plugin enabled or with the vanilla canvas.

![image](https://user-images.githubusercontent.com/1868974/132120942-3d19ec64-d323-485b-84aa-2b453bbee796.png)

**Comment by Adam-** [comment]
Does using reflection to `setActive(true)` on `com.apple.laf.ScreenPopupFactory` also work? It looks like it may since it hits the same `getHeavyWeightPopup`

```java
        final Popup popup;
        if (fIsActive) {
            popup = SwingAccessor.getPopupFactoryAccessor()
                    .getHeavyWeightPopup(this, comp, invoker, x, y);
        } else {
            popup = super.getPopup(comp, invoker, x, y);
        }
```

**Comment by Adam-** [comment]
fixed in 97095b41f8dc48d5cfa0064fc1d2b6847ad911aa on launcher 2.0+
